src/main/scala/is/hail/expr/AST.scala

```diff
@@ -741,21 +741,21 @@ case class Apply(posn: Position, fn: String, args: Array[AST]) extends AST(posn,
       FunctionRegistry.call(fn, args, args.map(_.`type`).toSeq)
   }
 
-  private val tryPrimOpConversion: IndexedSeq[IR] => Option[IR] = flatLift {
+  private def tryPrimOpConversion(types: IndexedSeq[Type]): IndexedSeq[IR] => Option[IR] = flatLift {
       case IndexedSeq(x) => for {
         op <- ir.UnaryOp.fromString.lift(fn)
-        t <- ir.UnaryOp.returnTypeOption(op, x.typ)
+        t <- ir.UnaryOp.returnTypeOption(op, types(0))
```

Comment:
I'm confused by this.  What was the problem?  The second (curried) argument (e.g. x) is no longer used.

Reply:
when the types are null because they haven't been inferred yet, this fails.

Reply:
Well, don't pass things in twice.  Either take types and return Option[IR] or return IndexedSeq[Types] => Option[IR], but not both.

Reply:
wait actually I don't understand. This is essentially what I'm doing in the function registry lookup (taking name and types and returning Option[Seq[IR] => IR]), except the Option is one level deeper (and `x` does get used---it's inside the constructed IR).

Reply:
Ah, I see, you're right.

Hmm.  This is quite confusing because types and the IR fields are analogous but they're treated completely differently.  Maybe zip them in the caller, then you have to deal with them the same way?